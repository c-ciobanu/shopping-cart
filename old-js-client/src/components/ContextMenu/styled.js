import styled from "styled-components";

export const StyledContainer = styled.div`
	position: relative;
	display: inline-block;
`;

export const StyledMenu = styled.ul`
	z-index: 1000;
	position: absolute;
	top: 0;
	right: 0;
	list-style: none;
	padding: 0;
	margin: 0;
	width: 150px;
	background-color: #fff;
	border: 1px solid lightgrey;
	outline: none;

	${({ isOpen }) =>
		!isOpen &&
		`
		border: 0;
	`}
`;

export const StyledMenuItem = styled.li`
	padding: 5px;
	cursor: pointer;

	${({ isHighlighted }) =>
		isHighlighted &&
		`
		background-color: lightgrey;
	`}
`;

export const StyledButton = styled.button`
	padding: 0;
	border: 0;
	outline: none;
	cursor: pointer;
`;
