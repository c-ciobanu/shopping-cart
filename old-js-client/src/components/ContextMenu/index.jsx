import React from "react";
import PropTypes from "prop-types";
import { useSelect } from "downshift";

import { ReactComponent as MoreIcon } from "assets/svg/more.svg";

import { StyledContainer, StyledMenu, StyledMenuItem, StyledButton } from "./styled";

function ContextMenu({ options }) {
	const { isOpen, getToggleButtonProps, getMenuProps, highlightedIndex, getItemProps } = useSelect({
		items: options
	});

	return (
		<StyledContainer>
			<StyledButton {...getToggleButtonProps()}>
				<MoreIcon />
			</StyledButton>

			<StyledMenu {...getMenuProps({ isOpen })}>
				{isOpen
					? options.map((item, index) => (
							<StyledMenuItem
								key={`li-${index}`}
								{...getItemProps({ item, index, isHighlighted: highlightedIndex === index, onClick: item.action })}
							>
								{item.text}
							</StyledMenuItem>
					  ))
					: null}
			</StyledMenu>
		</StyledContainer>
	);
}

ContextMenu.propTypes = {
	options: PropTypes.arrayOf(
		PropTypes.shape({
			text: PropTypes.string.isRequired,
			action: PropTypes.func.isRequired
		})
	).isRequired
};

export default ContextMenu;
