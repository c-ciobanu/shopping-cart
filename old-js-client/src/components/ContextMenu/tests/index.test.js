import React from "react";
import { render, fireEvent } from "@testing-library/react";

import ContextMenu from "../index";

describe("<ContextMenu />", () => {
	const callback = jest.fn();
	const options = [
		{ text: "Option 1", action: callback },
		{ text: "Option 2", action: callback },
		{ text: "Option 3", action: callback }
	];

	it("Expect to not log errors in console", () => {
		const spy = jest.spyOn(global.console, "error");
		render(<ContextMenu options={[]} />);
		expect(spy).not.toHaveBeenCalled();
	});

	it("Should correctly open and close menu", () => {
		const { container } = render(<ContextMenu options={options} />);

		const button = container.querySelector("button");
		const menu = container.querySelector("ul");
		let menuItems = container.querySelectorAll("li");

		expect(menuItems).toHaveLength(0);

		fireEvent.click(button);

		menuItems = container.querySelectorAll("li");

		expect(menuItems).toHaveLength(3);

		fireEvent.blur(menu);

		menuItems = container.querySelectorAll("li");

		expect(menuItems).toHaveLength(0);
	});

	it("Should correctly call callback on menu item click", () => {
		const { container, getByText } = render(<ContextMenu options={options} />);

		const button = container.querySelector("button");

		fireEvent.click(button);

		fireEvent.click(getByText(options[0].text));

		expect(callback).toHaveBeenCalled();
	});

	it("Should render and match the snapshot", () => {
		const {
			container: { firstChild }
		} = render(<ContextMenu options={[]} />);
		expect(firstChild).toMatchSnapshot();
	});
});
