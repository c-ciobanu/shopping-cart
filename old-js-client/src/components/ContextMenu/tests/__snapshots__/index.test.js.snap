// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContextMenu /> Should render and match the snapshot 1`] = `
<div
  class="sc-AykKC fnZNYC"
>
  <button
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-4-label downshift-4-toggle-button"
    class="sc-AykKF gvJIvP"
    id="downshift-4-toggle-button"
  >
    <svg>
      more.svg
    </svg>
  </button>
  <ul
    aria-labelledby="downshift-4-label"
    class="sc-AykKD hcHMvH"
    id="downshift-4-menu"
    role="listbox"
    tabindex="-1"
  />
</div>
`;
